Add CTest support for `mbed-drivers-ticker` greentea test

rwalton-arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
This PR adds CTest support for greentea tests, and enables it for the
`mbed-drivers-ticker` test.

This PR makes the following changes:
* Moves post-build functions out of root CMakeLists.txt to CMake module
* Adds a `BUILD_GREENTEA_TESTS` option to enable greentea tests
* Sets variables usually defined by external applications in root CMakeLists.txt
* Enable CTest for the `mbed-drivers-ticker` test, using `mbedhtrun` to run the test
* Adds an `MBED_HTRUN_ARGUMENTS` cache variable which is a list of options to forward to `mbedhtrun`

Checks have been added in `mbed_greentea.cmake` to keep the old flow working.
These checks should be removed once all tests are reworked.

Fixes #14852 
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
#14854 
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

rwalton-arm committed:
CMake: Add option to enable greentea tests

Add an option to enable the greentea tests independently from the unit
tests.

We can't just use the typical BUILD_TESTING option to enable greentea
tests. BUILD_TESTING enables unit tests and fetches googletest, which
are compiled for the host. Greentea tests are cross compiled and require
a toolchain file. For this reason we add a new option just to enable
greentea tests, preventing build failures triggered by the unit tests
and googletest.

rwalton-arm committed:
CMake: greentea: Set variables usually defined by external apps

We rely on external applications to set MBED_CONFIG_PATH and include
app.cmake. Mbed OS can't build for a target without those parameters
set. When building greentea tests there is no external application. So,
we define the appropriate settings in our root CMakeLists.txt for
greentea tests.

rwalton-arm committed:
CMake: greentea: Port mbed-drivers-ticker to CTest

Call add_test in mbed_greentea_add_test and specify mbedhtrun as the
command to run.

An MBED_HTRUN_ARGUMENTS variable has been added, which is a semicolon
separated list of arguments to forward to htrun. The user is required to
pass in any arguments mbedhtrun needs to communicate with the device
under test at CMake configuration time.

We automate passing some of the htrun arguments in CMake using data
provided by mbed_config.cmake or gathered from the CMake environment.
The following arguments are passed to htrun from CMake:

* -f/--image-path
* -e/--enum-host-tests
* -m/--micro
* -R/--reset-timeout

--baud-rate is also passed if the MBED_CONF_PLATFORM_STDIO_BAUD_RATE
config parameter is defined.

Temporary checks have been added to mbed_greentea_add_test to keep the
old flow working until we port all of the greentea tests to CTest. These
checks should be removed after we make all greentea tests runnable by
CTest.

rwalton-arm committed:
CMake: greentea: Use more specific name for baremetal option

Make it obvious the option is used for greentea tests.

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -5,6 +5,9 @@
 cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
 
+option(BUILD_TESTING "Run unit tests only." OFF)
+option(BUILD_GREENTEA_TESTS "Build greentea tests only." OFF)
```
Worth raising an error if both are enabled?

LDong-Arm replied:
Documentation in tools/cmake/README.md needs an update eventually, though we'll need a mixture of "old" and "new" methods until we've updated all tests.

rwalton-arm replied:
Maybe a documentation update after we've finished porting all the tests is the best thing. I had considered that when I raised the PR, so I linked to the documentation issue in the description.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
-option(MBED_TEST_BAREMETAL OFF)
+option(MBED_GREENTEA_TEST_BAREMETAL "Select baremetal greentea tests" OFF)
```
Please update tools/cmake/README.md

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -3,10 +3,17 @@
+# TODO: After we convert all greentea tests to use CTest, remove this code. We
+# define these parameters in mbed-os/CMakeLists.txt for greentea tests.
+if(NOT BUILD_GREENTEA_TESTS)
+    set(MBED_CONFIG_PATH ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "")
+    include(${CMAKE_CURRENT_LIST_DIR}/app.cmake)
+    set(MBED_ROOT ${CMAKE_CURRENT_LIST_DIR}/../.. CACHE INTERNAL "")
```
Could use `options()` instead of `CACHE`, though it's been like this before this PR.

rwalton-arm replied:
Options only typically allow boolean type ON or OFF values. We need string values for these variables so I think CACHE STRING is more idiomatic.

LDong-Arm replied:
Fair enough

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +81,17 @@ macro(mbed_greentea_add_test)
+            -p ${MBED_TARGET_SERIAL_PORT}
+            -t ${MBED_TARGET_SERIAL_NUM}
+            -d ${MBED_TARGET_MOUNT_POINT}
+            -e ${MBED_GREENTEA_HOST_TESTS_DIR}
```
Is it possible to make it optional? For example, many tests do not require any interaction with the host apart from the "start" signal and reporting of success/failure, in which case `default_auto.py` from `htrun` itself is used.

LDong-Arm reviewed (changes requested):
@rwalton-arm Thanks, this looks and works really well! I just added some comments.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +81,17 @@ macro(mbed_greentea_add_test)
+        NAME ${MBED_GREENTEA_TEST_NAME}
+        COMMAND mbedhtrun
+            -m ${MBED_TARGET}
+            -f ${CMAKE_CURRENT_BINARY_DIR}/${MBED_GREENTEA_TEST_NAME}.bin
```
```suggestion
            -f ${MBED_GREENTEA_TEST_NAME}.bin
```
Forgot to mention: This can be simplified because `${CMAKE_CURRENT_BINARY_DIR}` is the default working directory: https://cmake.org/cmake/help/latest/prop_test/WORKING_DIRECTORY.html

LDong-Arm replied:
Also, some targets have .hex images. We can use `${MBED_OUTPUT_EXT}` defined in `mbed_config.cmake`, and fall back to `bin` if it's not set or an empty string.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +81,17 @@ macro(mbed_greentea_add_test)
+        COMMAND mbedhtrun
+            -m ${MBED_TARGET}
+            -f ${CMAKE_CURRENT_BINARY_DIR}/${MBED_GREENTEA_TEST_NAME}.bin
+            -p ${MBED_TARGET_SERIAL_PORT}
```
This will default to 9600 baud but some targets have 115200 (e.g. ARM_MUSCA_S1). We can get it from MBED_CONF_PLATFORM_STDIO_BAUD_RATE.

htrun supports two ways to specify baud rate:
* `-p <port> --baud-rate=<baud>`
* `-p <port>:<baud>`

rwalton-arm replied:
I've added another commit to fetch the baud rate from the config. It's not pretty, because we don't define individual variables for the `MBED_CONF ...` macros. They're just string values appended to a single CMake list, so we have to scrape the baud rate from the list. Please review.

LDong-Arm reviewed (dismissed):
I've added a few more comments

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +81,17 @@ macro(mbed_greentea_add_test)
+            -m ${MBED_TARGET}
+            -f ${CMAKE_CURRENT_BINARY_DIR}/${MBED_GREENTEA_TEST_NAME}.bin
+            -p ${MBED_TARGET_SERIAL_PORT}
+            -t ${MBED_TARGET_SERIAL_NUM}
```
It looks like `-m ${MBED_TARGET}` and `-t ${MBED_TARGET_SERIAL_NUM}` are optional too. Given serial port and mount point, it's usually enough for htrun to run a test on a local board. I think the target name `-m` is useful with the `--grm` option to let RaaS know which target to allocation, and the serial number might be to do with `--copy=COPY_METHOD` and `--reset=FORCED_RESET_TYPE` which might involve some DAPLink specifics (I'm not certain though).

The idea is it'd be good to make some of them optional so users don't need to specify them all.

LDong-Arm replied:
Anyway, this looks good as an initial PR. Maybe we can fine tune htrun command line options when working on #14908 which involves passing no predefined serial port or mount point which are allocated by RaaS.

rwalton-arm replied:
The way CMake forwards optional arguments to CTest is through the [CMAKE_CTEST_ARGUMENTS](https://cmake.org/cmake/help/latest/variable/CMAKE_CTEST_ARGUMENTS.html) variable, which accepts a semicolon separated list of optional arguments. I was thinking it would be good to do something similar for `htrun`, i.e `MBED_HTRUN_OPTIONAL_ARGUMENTS`, and the user can pass any optional flags when they need to (whether that's automatically through use of `mbed-tools`, or using some other method). I agree it's probably better to implement the fine tuning of command line options in a follow up PR.

rwalton-arm replied:
Added `MBED_HTRUN_OPTIONAL_ARGUMENTS` and used it to make `HOST_TEST_DIRS` optional. We can follow the same pattern when we refine the `htrun` argument list, if we think it's a good enough approach.

LDong-Arm replied:
Looks good to me

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -1,12 +1,19 @@
-set(MBED_CONFIG_PATH ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "")
+set(MBED_TARGET_SERIAL_PORT "" CACHE STRING "Serial port of the DUT")
+set(MBED_TARGET_SERIAL_NUM "" CACHE STRING "USB Serial number of the DUT")
+set(MBED_TARGET_MOUNT_POINT "" CACHE STRING "Mount point for the DUT")
```
It's a limitation of CTest that we can't pass any options at run time 😢: https://gitlab.kitware.com/cmake/cmake/-/issues/20470
If a user switches to a different board or connects a board after compilation, they need to rerun cmake to use their device. Though this won't require recompilation.

rwalton-arm replied:
Correct. Users will need to run the CMake config and generate step again if they want to change the cache values.

LDong-Arm reviewed (commented):
Verified on K64F, after rebasing the PR to have the greentea-client I/O fix.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +82,28 @@ macro(mbed_greentea_add_test)
+    if(NOT ${MBED_OUTPUT_EXT} STREQUAL "")
+        set(MBED_GREENTEA_TEST_IMAGE_NAME "${MBED_GREENTEA_TEST_NAME}.${MBED_OUTPUT_EXT}" CACHE INTERNAL "")
+    else()
+        set(MBED_GREENTEA_TEST_IMAGE_NAME "${MBED_GREENTEA_TEST_NAME}.bin" CACHE INTERNAL "")
```
Why is it a cache variable? I don't think there's any chance for the user to override the test image name, given that we'll eventually build all tests from the top of mbed-os directly and there will be more than one test names. Or is CACHE for another purpose here?

rwalton-arm replied:
I've removed CACHE as it doesn't really make sense for the user to set this variable.

LDong-Arm reviewed (commented):
Thanks for the update, I just added one question.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +81,37 @@ macro(mbed_greentea_add_test)
+    endif()
+
+    if(DEFINED MBED_TARGET)
+        list(APPEND MBED_HTRUN_OPTIONAL_ARGUMENTS "-m;${MBED_TARGET}")
```
I think
```cmake
list(APPEND MBED_HTRUN_OPTIONAL_ARGUMENTS -m ${MBED_TARGET})
```
has the same effect as lists are semicolon separated. From CMake's point of view, `-m` and `${MBED_TARGET}` are just two different things it passes to htrun (even though we know they are related).

And we may always assume MBED_TARGET to be defined as greentea tests run on device and require mbed_config.cmake.

rwalton-arm replied:
It does have the same effect, so it's a style point. I'm fine with the way it is now.

I also don't like assuming MBED_TARGET is always defined, as I think we want to break any hard dependencies on config going forward. We assume MBED_CONFIG_DEFINITIONS for extracting the baud rate, but we also check if we've actually found a valid value before passing it to htrun. So I'd prefer to go with the more defensive approach here.

LDong-Arm replied:
This makes sense

Patater commented on CMakeLists.txt:
```diff
@@ -19,14 +22,16 @@ list(APPEND CMAKE_MODULE_PATH
 endif()
 
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME AND BUILD_GREENTEA_TESTS)
+    enable_testing()
```
Do we not want to `include(CTest)` if only `BUILD_GREENTEA_TESTS` is set (and not `BUILD_TESTING`)?

rwalton-arm replied:
We could `include(CTest)`, it defines some extra options that could be useful if we wanted to use CDash. 

However, we need to first set `BUILD_TESTING = ON` here. This is because we define the `BUILD_TESTING` option at the top of our root CMakeLists.txt and set it `OFF`. So when CTest.cmake defines `BUILD_TESTING` again, and gives it a default `ON` value, it takes no effect, as we have already created a cache entry with the same name. This effectively tells CMake to ignore CTest.cmake's `option` call (see https://cmake.org/cmake/help/latest/command/option.html).  Ultimately that means `enable_testing` isn't called when we `include(CTest)` because in CTest.cmake `enable_testing` is protected by `if(BUILD_TESTING)`.

I think the correct change here is to remove our definition of `BUILD_TESTING` and add the following to our top level CMakeLists.txt:
```
if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
    include(CTest)
endif()
```
Then if we're building mbed-os as the top level project, `BUILD_TESTING` will be `ON` by default (as set in CTest.cmake). This is how we set up greentea-client and is the standard way of doing things with CTest. I'll try making the change and see if I run into any other problems.

Patater commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -3,10 +3,17 @@
-set(MBED_CONFIG_PATH ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "")
+set(MBED_TARGET_SERIAL_PORT "" CACHE STRING "Serial port of the DUT")
+set(MBED_TARGET_MOUNT_POINT "" CACHE STRING "Mount point for the DUT")
+set(MBED_HTRUN_OPTIONAL_ARGUMENTS "" CACHE STRING "Optional argument list to forward to htrun.")
```
We discussed potentially having just a single list of arguments instead of a handful of individual arguments and a list to append. Do we want to go with a single list from the beginning? How does the `cmake` commandline command to pass the arguments in look?

rwalton-arm replied:
We're going with a single list from the beginning, I'm pushing the changes for that soon.

LDong-Arm reviewed (approved):
Looks even better now.

I verified:
* local board: `-DMBED_HTRUN_ARGUMENTS='-d;/Volumes/DAPLINK/;-p;/dev/tty.usbmodem141102'`
* RaaS: `-DMBED_HTRUN_ARGUMENTS='--grm;raas_client:http://<address>'` (@ARMmbed/mbed-os-test: FYI)

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -19,12 +21,14 @@ list(APPEND CMAKE_MODULE_PATH
 
 add_subdirectory(extern)
 
-option(BUILD_TESTING "Run unit tests only." OFF)
```
This line was here for a reason. Without it, `include(CTest)` sets `BUILD_TESTING` to `ON` and breaks `mbed-tools compile` of applications

rwalton-arm replied:
Good catch. This is because our examples define the application project after adding the mbed-os subdirectory. So the mbed-os project is considered the "current project" when this file is parsed. If I make the following change to e.g mbed-os-example-blinky's CMakeLists.txt it compiles fine here:

```

 include(${MBED_PATH}/tools/cmake/app.cmake)

+project(${APP_TARGET})
+
 add_subdirectory(${MBED_PATH})

 add_executable(${APP_TARGET})

 mbed_configure_app_target(${APP_TARGET})

-project(${APP_TARGET})
```

I think the correct thing would be to fix the examples so they define their project before adding the mbed-os subdirectory.

LDong-Arm replied:
That makes sense. Shall we fix all examples as a priority, then get this PR merged? Or merge this with the old workaround, fix examples at some point, then come back to remove the workaround?

LDong-Arm reviewed (dismissed):
Found an issue

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -70,6 +79,35 @@ macro(mbed_greentea_add_test)
+    if(DEFINED MBED_TARGET)
+        list(APPEND MBED_HTRUN_ARGUMENTS "-m;${MBED_TARGET}")
+    endif()
+
```
Now we need to get https://github.com/ARMmbed/mbed-tools/pull/301 in and add
```suggestion

    if(NOT "${MBED_GREENTEA_TEST_RESET_TIMEOUT}" STREQUAL "")
        list(APPEND MBED_HTRUN_ARGUMENTS "-R;${MBED_GREENTEA_TEST_RESET_TIMEOUT}")
    endif()
```

rwalton-arm replied:
Added.

LDong-Arm reviewed (dismissed):
@rwalton-arm suggests raising the post-build fix as a follow-up PR. Shall we get this PR merged first?

rwalton-arm committed:
ci: Add github action to build greentea tests with cmake

Build for two targets, one with baremetal profile and one full profile.

rwalton-arm committed:
CMake: Use CMAKE_MODULE_PATH to locate nuvoton post build hook

We were previously relying on the global MBED_PATH variable to provide
the root for the include path to the post build hooks. We can't
guarantee that MBED_PATH will be set by any application building us, so
use CMAKE_MODULE_PATH (which we always set) like we do for the other
post-build hooks.

rwalton-arm committed:
targets: Add "bare-metal" to supported_application_profiles

We weren't setting the "supported_application_profiles" correctly for
some baremetal supported targets. This didn't cause build failures with
mbed-cli 1 as it just seems to ignore this config parameter. In our CMake
build system we have added a hard dependency on the
supported_application_profiles; we fail the build if the profile we're
trying to build isn't in the list.

hazzlim commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -30,19 +35,23 @@ macro(mbed_greentea_add_test)
         TEST_INCLUDE_DIRS
         TEST_SOURCES
         TEST_REQUIRED_LIBS
+        HOST_TESTS_DIR
```
Does this argument need to be added to the mbed_greentea_add_test documentation comment?

rwalton-arm replied:
Good spot. Yes, I'll add it to the comment.

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,8 @@
 
```
In commit "CMake: Add option to enable greentea tests ", please explain why "Add an option to enable the greentea tests independently from the unit tests." is important to have. Why is doing it this way helpful? What are the alternatives and why do we prefer this approach?

Patater commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -30,19 +37,23 @@ macro(mbed_greentea_add_test)
-    add_executable(${MBED_GREENTEA_TEST_NAME})
+    # TODO: After we convert all greentea tests to use CTest, remove this
+    # add_subdirectory call. We will attach the tests to the mbed-os project,
+    # rather than creating a new project for each test that depends on mbed-os.
```
Is there a tracking issue for this TODO which we can link to here?

rwalton-arm replied:
https://github.com/ARMmbed/mbed-os/issues/14991

Patater commented on .github/workflows/greentea_cmake.yml:
```diff
@@ -0,0 +1,30 @@
+
+      - name: Install the latest mbed-tools
+        run: |
+            pip3 install --upgrade mbed-tools
```
Do we need to upgrade mbed-tools or can we trust the "latest" mbed-os-env image has a working version?

LDong-Arm replied:
I added this because mbed-tools in the Docker image was outdated and GitHub Actions failed. Discussion https://github.com/ARMmbed/mbed-os/pull/14892#issuecomment-891006417 and https://github.com/ARMmbed/mbed-os/pull/14892#issuecomment-891122643. This is to ensure we have the latest released version.

LDong-Arm replied:
It looks like the latest Docker image is still from a month ago: https://hub.docker.com/r/mbedos/mbed-os-env/tags?page=1&ordering=last_updated

rwalton-arm replied:
I suggest leaving this in because the mbed-os-env doesn't have a working version.

Patater replied:
Could we file an issue to remove this? mbed-os-env really ought to have a working version

rwalton-arm replied:
https://github.com/ARMmbed/mbed-os/issues/14992

rwalton-arm replied:
We were using the wrong docker image (see Saheer's comment below) - now fixed.

Patater reviewed (dismissed):
LGTM

Let's get the final CI runs in and merge this finally.

saheerb commented on .github/workflows/greentea_cmake.yml:
```diff
@@ -0,0 +1,30 @@
+jobs:
+  build-greentea-cmake:
+    runs-on: ubuntu-latest
+    container: mbedos/mbed-os-env:latest
```
mbedos/mbed-os-env in dockerhub will be deprecated soon.

Please point to 
`ghcr.io/armmbed/mbed-os-env:master-latest`

rwalton-arm replied:
We also need to change https://github.com/ARMmbed/mbed-os/blob/master/.github/workflows/test_building_multiple_executables.yml

Patater replied:
Good find. This was the root cause of our having out of date tools and needing to upgrade them manually. We can remove that by using the right Docker image.

rwalton-arm committed:
ci: Use new mbed-os-env docker image in "multiple executables" test

The github workflow to test building multiple executables was using the
soon to be deprecated docker image "mbedos/mbed-os-env:latest". This
commit points it at the new docker image on ghcr.io.
